README
p-boot - pico sized bootloader for PinePhone
--------------------------------------------

Official website: https://xnux.eu/p-boot/

This is a no nonsense bootloader for extremely fast and flexible booting
of PinePhone. p-boot is partly based on some borrowed U-Boot and Linux code.

Some benefits over U-Boot:

- It's very fast (30-60ms boot times + load times for images)
  - Uses MMC DMA, eMMC DDR, and CPU data cache
  - Only minimal necessary boot steps are performed in the most straightforward manner
- It's fail safe and easily debuggable
  - It will power down the device in case of bootloader stage boot failure,
    instead of silently draining the battery
  - Boot status is indicated via LED
    - green  = in bootloader
    - red    = just before passing control to ATF/Linux
    - no led = powered off
  - Bootloader failure is indicated by a 5-bit binary blinking pattern
    (serial port not needed for quick debugging)
  - Bootloader log is stored to DTB, and is visible from Linux via
    `/sys/firmware/devicetree/base/p-boot/log` file
  - Boot log can be printed to the serial port (this will slow down boot)
- It has flexible boot source options
  - Supports multiple boot configurations on SD and eMMC
  - Ability to load bootloader from SD card, but read boot data from eMMC (which
    allows 85MiB/s image load speeds and avoids 1s LED flash activation and 1s boot
    delay waiting for BROM to load bootlader from eMMC)
  - List of available boot configurations is passed to Linux via
    `/sys/firmware/devicetree/base/p-boot/configs` file, so it's easy to make
    a Linux based boot menu.
  - Boot configuration selection via volume keys pressed during boot
    (up to 3 different boot configurations)
  - Allow for default boot config override via a RTC data register (you can reboot
    from one OS to another without user intervention)
  - Support for one time reboot to other OS, or "persistent" change to other
    boot configuration
- It has access to PMIC over RSB (p-boot can access and configure the PMIC)
  - Easy to select alternate boot configuration in case of low battery charge level
  - p-boot configures the PMIC and can react to various boot reasons indicated
    by PMIC
- Single DT blob used by p-boot, ATF and Linux (no U-Boot specific DT blob needed)
- Configure a stable WiFi MAC address in FDT based on SoC ID
- A lot of free space for extensions (p-boot size limit is 32KiB, current
  p-boot size is 31KiB with display support enabled and 25KiB with display support
  disabled)
- Easy to understand and customize

Enjoy!

    Ondřej Jirman (aka "megi")


Future possibilities
--------------------

See TODO for some possible/planned features.


Building
--------

1) Get aarch64-linux-musl cross-compiler and set path to it in config.ini
   (you can use config.ini.sample as a starting point).

2) Then run:

  php configure.php
  ninja

3) As a result you'll get a bunch of p-boot*.bin variants in .build/ dir,
   named p-boot*.bin. See "Variants" section bellow for description
   of each variant.


Building (without PHP)
----------------------

1) Get aarch64-linux-musl cross-compiler and set path to it in
   build/ninja.build by modifying aarch64_prefix.

2) Go to build and run:

  ninja

3) Same as above.


Installation summary / key points
---------------------------------

- PinePhone looks for bootloader 8KiB from the start of SD or eMMC block
  devices, in that order. p-boot.bin needs to be flashed there.
- p-boot doesn't understand traditional filesystems, it uses its own
  very simple boot filesystem format, and comes with a filesystem
  formating utility called p-boot-conf, that is used to create it.
- Once flashed to the boot sector, p-boot will look for the boot
  filesystem in the DOS MBR style partition table (it doesn't understand
  any other partition format).
- So all you need to have is:
  - DOS MBR partition table
  - A boot partition for p-boot. You can re-use a boot partition
    that most distributions create for files in /boot, just copy files from
    /boot to your root partition, unmount /boot, and re-format the boot
    partition using p-boot-conf. You can copy the the original files
    from boot partition back to /boot directory on the root partition,
    so that your distro will not break and keep updating them. Just
    make sure your distro will not overwrite a boot sector during updates
    with U-Boot. (perhaps by uninstalling a u-boot package)


Installing p-boot to the "boot sector"
--------------------------------------

You must install p-boot by copying the contents of p-boot.bin to the SD card
and/or eMMC starting from 8KiB offset from the start of the block device:

  dd if=p-boot.bin of=/dev/mmcblk2 bs=1024 seek=8
  dd if=p-boot.bin of=/dev/mmcblk0 bs=1024 seek=8


ATF / SCP
---------

You'll need a patched ATF that works with p-boot and a SCP firmware build.

You can build these using instructions in the fw/README or use a pre-built
binary from dist/fw.bin (contains combined ATF and SCP binaries).


Boot partition
--------------

p-boot needs to find a specialized boot filesystem on one of the primary MBR
partitions on eMMC or SD card (in that order). p-boot will only search
partitions of type 0x83 (Linux) and it will give priority to partitions
with a boot flag set.

Boot partition contains a table of up to 33 boot configurations. Boot
configuration is a specific set of ATF/SCP firmware, DTB, Linux, and initramfs
images + Linux command line arguments (bootargs).

You can use p-boot-conf tool to create a boot filesystem. You can find this
tool pre-built in the dist/ folder.

This tool reads a 'boot.conf' file from a specified configuration directory
and creates a specialized filesystem on the boot partition based on the
content of these files. You'll need to run this tool each time you'll update
the configuration, or one of the kernel images, to regenerate the boot
filesystem.

Each configuration file has the following format:

  key1=value
  key2=value
  ...

For example:

  no=1
  dtb=board.dtb
  atf=fw.bin
  linux=Image
  initramfs=initramfs.img
  bootargs=console=ttyS0,115200 console=tty1 root=/dev/mmcblk0p2 rootfstype=f2fs rw rootwait panic=3

Meanings of configuration options:

  - no: Configuration index (0-32)
  - dtb: path to the pinephone's DTB file
  - atf: path to the ATF binary or to the combined ATF+SCP binary (see dist/fw.bin)
  - linux: path to the Linux 'Image' file
  - initramfs: path to initramfs archive
  - bootargs: Linux kernel boot arguments to be passed to the kernel

All options are required except for initramfs. All paths are relative to the
configuration directory.

After preparing the configuration files, and collecting the required binary files
in the configuration directory, run:

  p-boot-conf $conf_dir $boot_partition_block_device

Once done, you can reboot the PinePhone to check that p-boot works. Pre-built
dist/p-boot-conf is meant for running on PinePhone itself. If you need a build
of this tool for another architecture, just run `gcc -o p-boot-conf-native conf.c`
in the `src/` directory.


GUI variant of p-boot
---------------------

GUI variant of p-boot looks for certain hardcoded files in the boot filesystem:

- off.argb     - power off splashscreen
- pboot2.argb  - default p-boot splashscreen

These files can be found in the `example/` configuration, or generated
from source files in the `theme/` directory.


Runtime behavior
----------------

(this is slightly outdated, check main.c for actual behavior)

During powerup or reboot, p-boot will check several places in its decision
making process for selecting the desired boot configuration:

- boot filesystem superblock (32bit BE value at 12 byte offset from the
                              start of boot partition)
- RTC data register (address 0x01f00100)
- PMIC data register (address 0x04)
- Status of the volume keys:
  - VOLUME_UP - no=1
  - VOLUME_DOWN - no=0

Default value is stored in the boot filesystem's superblock. (at offset 12,
4-byte BE value). The value is unchanged by p-boot-conf.

Values in volatile registers are reset to 0 during boot if bit 7 of the value
is set. If the register contains value 0, the value is ignored and doesn't
affect boot configuration selection process. Values other than 0 are
decremented by 1 and the result is used to select the boot configuration.
For example register value 0x03 will select boot configuration 2. Register
value 0x82 will select boot configuration 1 and the register will be
reset to 0 during boot.


Boot process of p-boot is as follows: (also see src/main.c)

- sample the time of the start of the boot
- turn on a green LED
- init various A64 clocks
- enable lradc (to get a sample from volume keys ADC, later on)
- (optional) init debug UART
- init RSB connection to PMIC
- configure PMIC
  - (optional) print PMIC status
  - shorten POK reaction time
  - increase VBUS current limit to 2A and enable charger detection
  - increase CPU voltage to accommodate switch to 1.152GHz later on
- init DRAM
  - setup p-boot heap space for malloc() at 256MiB from end of DRAM
  - enable instruction cache
  - enable MMU/setup CPU data cache
- setup SoC buses
- enable LCD power early so that controller initializes before
  kernel runs, and kernel LCD probe can avoid delays
- disable battery temperature sensor
- program the battery OCV curve for better capacity reporting
- increase CPU clock speed to 1152MHz
- initialize eMMC
  - search for boot partition and load bootfs superblock/conf table
  - if not found repeat the same process on the SD card
- select configuration from the boot table
  - read default boot config selection index from the table
  - apply value from the RTC register
  - apply value from the PMIC register
  - if VOLUME_UP is pressed, force bootsel index = 1
  - if VOLUME_DOWN is pressed, force bootsel index = 0
  - go through the table and load images from the table to:
    - Linux image to 0x40080000
    - initramfs to 0x4fe00000
    - DTB (FDT blob) to 0x4a000000
    - ATF+SCP to 0x44000
- check validity of FDT blob, and in FDT:
  - set /chosen/bootargs to kernel command line
    - append bootdev=emmc or bootdev=sd based on the source of the bootfs
  - calculate MAC address for WiFi from SoC ID and patch it to FDT
  - add DRAM address/size info
  - add initramfs info
- turn off green and turn on red LED
- jump to ATF
  - ATF jumps to Linux image while switching to EL2
  - (alternatively) after return from ATF, jump to Linux image while switching to EL2


Debugging
---------

p-boot will try to fail gracefully in the event of boot failure. This means
that it will try its best to power down the SoC. p-boot also enables the red
LED as early as possible so that the user is informed in case the boot process
hangs that the device is on and drawing power from the battery.

In case of a bootloader panic, p-boot will first blink the red led quickly for
a second and then you'll see 5 flashes of the green LED followed by device poweroff.
The 5 green flashes represent a 5-bit binary number in MSB first bit order.
Long flash represents 1 and short flash represents 0.

You can use this to get the reason for boot failure without the need to pull
out the serial cable. Simply translate the binary number to decimal and search
p-boot code for the panic call with this number. For example, if you get
a flash pattern ". - . - ." -> 0b01010 -> 10:

  grep -R 'panic(10' src

src/main.c:		panic(10, "BOOTFS not found");

That is bootfs was not found. :)


Variants
--------

p-boot can be compiled with several options that improve boot speed,
or reduce size or both.

  p-boot.bin        - GUI variant of p-boot
  p-boot-serial.bin - non-GUI variant, logs to serial console and FDT
  p-boot-tiny.bin   - non-GUI variant, doesn't log anyhting at all,
                      and doesn't store log messages in the binary,
                      so it saves about 5 KiB of space for more code

This is a typical boot log from the p-boot-serial.bin:

% cat /sys/firmware/devicetree/base/p-boot/log

p-boot (version 547db8d-dirty built 2020-05-20 22:55)
384 us: PMIC ready
3492 us: DRAM: 2048 MiB
SoC ID: 92c059ba:14104620:78968308:101a0950
Board rev: 1.2
Boot Source: ff
34065 us: eMMC ready
Trying boot partition located at 0x100000(255 MiB) (part 0)
Bootsel override via RTC data[0] to 5
35500 us: Booting configuration 5 (xnux:Tablet UI test (eMMC))
41228 us: IMG[A]: 0x476e400(61368 B) -> 0x44000 (10466 KiB/s)
41856 us: IMG[D]: 0x477d400(37399 B) -> 0x4a000000 (58342 KiB/s)
196896 us: IMG[L]: 0x4786800(14002184 B) -> 0x40080000 (88197 KiB/s)
Model: Pine64 PinePhone (1.2)
WiFi MAC: 02:ba:7c:9c:cc:78


Bugs and support
----------------

Please report bugs via email:

    p-boot@xnux.eu


Development
-----------

Project status: p-boot is maintained and actively developed by the original
author.

You can join development of p-boot by sending questions, ideas, bugreports
and patches to the above mailing address.

GIT repository is available at: https://megous.com/git/p-boot. 


Support the project
-------------------

p-boot is free and opesource software released under GPLv3. It is an independent
personal project not backed by any comapny or commercial interests. You can
contribute by writing code, helping improve documentation, spreading the word,
and/or by donating at:

  https://xnux.eu/contribute.html



NEWS
p-boot 0.9 (2020-09-01)
-----------------------

- Cleaned up codebase significantly
- Added display support
- Added boot menu and virtual console support



src/conf.c
/**
 * p-boot - pico sized bootloader
 *
 * Copyright (C) 2020  Ondřej Jirman <megi@xff.cz>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <assert.h>
#include <fcntl.h>
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <stdbool.h>
#include <stdint.h>
#include <stddef.h>
#include <endian.h>
#include <string.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <dirent.h>

#include "bootfs.h"
#ifndef PATH_MAX
#define PATH_MAX 1024
#endif

/*
 * A tool for creating a bootfs filesystem.
 *
 * This is a simple read-only filesystem for p-boot, that contains file data in
 * continuous segments that can be read very quickly without any indirection.
 *
 * Metadata is stored in the first 64 KiB in 32 2 KiB blocks. There are 3 types
 * of metadata blocks:
 * - superblock: contains information about the filesystem as a whole (block 0)
 * - bconf: contains one boot configuration (blocks 1-31)
 * - files: contains a list of 51 file nodes (name + data offset/length) (blocks 1-31)
 *          filename size limit is 31 characters
 */

struct data {
	char path[PATH_MAX];
	int fd;
	uint32_t offset;
	uint32_t size;
	struct data* next;
};

struct file {
	char name[32];
	struct data* data;
};

struct bconf_image {
	uint32_t type;
	struct data* data;
	struct bconf_image* next;
};

struct bconf {
	int used;
	int index;
	char path[1024];
	char name[1024];
	char bootargs[4096];
	struct bconf_image* images;
};

static char* device_id;
static struct data* data_list;
static struct bconf confs[32];
static int n_files;
static struct file files[400];

// {{{ Parse conf file

static struct data* data_add_file(const char* path)
{
	struct data* d, *last_d;
	char rpath[PATH_MAX];

	if (!realpath(path, rpath)) {
		printf("ERROR: Can't resolve path '%s' (%s)", path, strerror(errno));
		exit(1);
	}

	for (d = data_list, last_d = d; d; last_d = d, d = d->next) {
		if (!strcmp(rpath, d->path))
			return d;
	}

	d = malloc(sizeof *d);
	assert(d != NULL);
	memset(d, 0, sizeof *d);

	int fd = open(rpath, O_RDONLY);
	if (fd < 0) {
		printf("ERROR: Can't open file '%s' (%s)", path, strerror(errno));
		exit(1);
	}

	//XXX: maybe read file contents and avoid dups based on content too

	snprintf(d->path, sizeof d->path, "%s", rpath);
        d->fd = fd;

	if (last_d)
		last_d->next = d;
	else
		data_list = d;

	return d;
}

static const struct image_type {
	const char* conf_var;
	char type;
	bool optional;
} image_types[] = {
	{ "linux",     'L', },
	{ "initramfs", 'I', true },
	{ "atf",       'A', },
	{ "dtb",       'D', },
	{ "splash",    'S', true },
};

static void complete_conf(struct bconf* c)
{
	if (confs[c->index].used) {
		printf("ERROR: %s: Configuration slot no=%d is already taken by '%s'\n", c->path, c->index, confs[c->index].path);
		exit(1);
	}

	if (!c->name[0]) {
		printf("ERROR: %s: Configuration slot no=%d is missing a name\n", c->path, c->index);
		exit(1);
	}

	for (int i = 0; i < sizeof(image_types) / sizeof(image_types[0]); i++) {
		if (image_types[i].optional)
			continue;

		struct bconf_image* im;
                for (im = c->images; im; im = im->next) {
			if (im->type == image_types[i].type)
				goto found;
		}

		printf("ERROR: %s: Configuration slot no=%d is missing '%s' image\n", c->path, c->index, image_types[i].conf_var);
		exit(1);
found:;
	}

	c->used = 1;
	confs[c->index] = *c;
}

static bool parse_conf(const char* conf_dir, const char* conf_filename)
{
	struct bconf conf_tpl = {};

	snprintf(conf_tpl.path, sizeof conf_tpl.path, "%s/%s", conf_dir, conf_filename);

	FILE* f = fopen(conf_tpl.path, "r");
	if (!f) {
		printf("ERROR: can't open %s", conf_tpl.path);
		exit(1);
	}

	char line[4096];
	struct bconf conf = conf_tpl;
	bool conf_started = false;
	int line_no = 0;

	while (fgets(line, sizeof line, f)) {
		char* end = line + strlen(line);
		while (--end >= line && *end == ' ' || *end == '\t' || *end == '\n')
			*end = 0;
                char* name = line;
                line_no++;

		while (*name == ' ' || *name == '\t')
			name++;

		if (*name == '#')
			continue;
		if (*name == '\0')
			continue;

		char* eq = strchr(name, '=');
		if (!eq) {
			printf("WARNING: %s[%d]: Skipping invalid line (%s)", conf.path, line_no, line);
			continue;
		}

		char* val = eq + 1;

		// trim name from the end
		do {
			*eq-- = 0;
		} while (eq >= name && (*eq == ' ' || *eq == '\t'));

		// trim value from the start
		while (*val == ' ' || *val == '\t')
			val++;

		if (val) {
			if (!strcmp(name, "device_id")) {
				if (device_id) {
					printf("ERROR: %s[%d]: multiple device_id are not allowed", conf.path, line_no);
					exit(1);
				}

				if (strlen(val) > 31) {
					printf("ERROR: %s[%d]: device_id is too long (max 31 chars)", conf.path, line_no);
					exit(1);
				}

				device_id = strdup(val);
				continue;
			}

			// got a reasonably valid line
			if (!strcmp(name, "no")) {
				if (conf_started)
					complete_conf(&conf);

				conf = conf_tpl;
				conf.index = atoi(val);
				if (conf.index < 0 || conf.index >= 32) {
					printf("ERROR: %s[%d]: no out of range (is %d, must be 0-32)", conf.path, line_no, conf.index);
					exit(1);
				}

				conf_started = true;
				continue;
			}

			if (!conf_started) {
				printf("ERROR: %s[%d]: Config has to start with no=[idx]", conf.path, line_no);
				exit(1);
			}

			if (!strcmp(name, "name"))
				snprintf(conf.name, sizeof conf.name, "%s", val);

			if (!strcmp(name, "bootargs"))
				snprintf(conf.bootargs, sizeof conf.bootargs, "%s", val);

			for (int i = 0; i < sizeof(image_types) / sizeof(image_types[0]); i++) {
				if (strcmp(name, image_types[i].conf_var))
					continue;

				char path[PATH_MAX];
				snprintf(path, sizeof path, "%s/%s", conf_dir, val);
				struct data* d = data_add_file(path);

				struct bconf_image* im = malloc(sizeof *im), *imi, *imi_last;
				assert(im != NULL);
				memset(im, 0, sizeof *im);

				for (imi = conf.images, imi_last = imi; imi; imi_last = imi, imi = imi->next) {
					if (imi->type == image_types[i].type) {
						printf("ERROR: %s[%d]: Image '%s' is already set for no=%d", conf.path, line_no, image_types[i].conf_var, conf.index);
						exit(1);
					}
				}

				im->type = image_types[i].type;
				im->data = d;

				if (imi_last)
					imi_last->next = im;
				else
					conf.images = im;
			}
		}
	}

	if (conf_started)
		complete_conf(&conf);

	fclose(f);
	return true;
}

static void include_files(const char* dir)
{
	DIR* d = opendir(dir);
	if (!d) {
		printf("ERROR: Can't open directory '%s'\n", dir);
		exit(1);
	}

	struct dirent *e;
	while (1) {
		errno = 0;
		e = readdir(d);
		if (!e) {
			if (errno) {
				printf("ERROR: failed to read directory '%s'\n", dir);
				exit(1);
			}

			break;
		}

		char path[PATH_MAX];
		snprintf(path, sizeof path, "%s/%s", dir, e->d_name);
		struct stat st;
		int ret = stat(path, &st);
		if (ret) {
			printf("ERROR: failed to stat %s\n", path);
			exit(1);
		}

		if (S_ISREG(st.st_mode)) {
			if (strlen(e->d_name) > 31) {
				printf("ERROR: File name too long '%s', max is 31 chars\n", path);
				exit(1);
			}

			if (n_files >= sizeof(files) / sizeof(files[0])) {
				printf("ERROR: Too many files\n", path);
				exit(1);
			}

			struct data* d = data_add_file(path);
			struct file* f = &files[n_files++];

			f->data = d;
			snprintf(f->name, sizeof f->name, "%s", e->d_name);
		}
	}

	closedir(d);
}

// }}}
// {{{ Write filesystem

void write_checked(int fd, void* buf, size_t len)
{
	ssize_t ret = write(fd, buf, len);
	if (ret < 0 || ret != len) {
		printf("ERROR: failed writing bootfs!!! %s\n", strerror(errno));
		exit(1);
	}
}

off_t lseek_checked(int fd, off_t off)
{
	off_t off_out = lseek(fd, off, SEEK_SET);
	if (off_out < 0) {
		printf("ERROR: failed writing bootfs!!! %s\n", strerror(errno));
		exit(1);
	}

	return off_out;
}

size_t write_fd_checked(int dest_fd, int src_fd)
{
	size_t len = 0;

	while (1) {
		char buf[1024*128];

		ssize_t read_b = read(src_fd, buf, sizeof buf);
		if (read_b < 0) {
			printf("ERROR: failed writing bootfs!!! %s\n", strerror(errno));
			exit(1);
		} else if (read_b == 0) {
			break;
		}

		ssize_t wr_b = write(dest_fd, buf, read_b);
		if (read_b != wr_b) {
			printf("ERROR: failed writing bootfs!!! %s\n", strerror(errno));
			exit(1);
		}

		len += wr_b;
	}

	return len;
}

static void usage(const char* msg)
{
	printf("ERROR: %s\n", msg);
	printf("Usage: p-boot-conf <conf-dir> <blk-dev>\n\n");
	printf("Example: p-boot-conf /boot /dev/mmclbk1p1\n");
	exit(1);
}

int main(int ac, char* av[])
{
	const char* conf_dir;
	const char* blk_dev;

	if (ac != 3)
		usage("mising options");

	conf_dir = av[1];
	blk_dev = av[2];

	parse_conf(conf_dir, "boot.conf");

	char path[PATH_MAX];
	snprintf(path, sizeof path, "%s/files", conf_dir);
	include_files(path);

	/* open bootfs partition block device */
	int fd = open(blk_dev, O_RDWR | O_CREAT | O_TRUNC, 0666);
	if (fd < 0) {
		perror("can't open block device\n");
		exit(1);
	}

	lseek_checked(fd, 0);

	struct bootfs_sb sb = {
		.magic = ":BOOTFS:",
		.version = htobe32(1),
	};
	
	if (device_id)
		snprintf(sb.device_id, sizeof sb.device_id, "%s", device_id);

	write_checked(fd, &sb, sizeof sb);

	off_t off_c = 2048;
	off_t off_i = 2048 * 33;
	int files_written = 0;

        // write data images
	printf("Data space:\n\n");
	for (struct data* d = data_list; d; d = d->next) {
		lseek_checked(fd, off_i);
		lseek_checked(d->fd, 0);

		d->offset = off_i;
		d->size = write_fd_checked(fd, d->fd);

		off_i += d->size;
		if (off_i % 512)
			off_i += (512 - off_i % 512);

		printf("    %08x-%08x: %s (size %u KiB)\n", d->offset, d->offset + d->size, d->path, d->size / 1024);
	}

	printf("\nBoot configurations:\n\n");
	for (int i = 0; i < 32; i++) {
		if (confs[i].used) {
			struct bootfs_conf bc = {
				.magic = ":BFCONF:",
			};

			printf("no=%d (%s)\n\n", confs[i].index, confs[i].name);
			printf("  %s\n\n", confs[i].bootargs);

			snprintf(bc.name, sizeof bc.name, "%s", confs[i].name);
			snprintf(bc.boot_args, sizeof bc.boot_args, "%s", confs[i].bootargs);

			int n_imgs = 0;
			for (struct bconf_image* im = confs[i].images; im; im = im->next) {
				bc.images[n_imgs].type = htobe32(im->type);
				bc.images[n_imgs].data_off = htobe32(im->data->offset);
				bc.images[n_imgs++].data_len = htobe32(im->data->size);

				printf("  %c %08x-%08x %s\n", im->type, im->data->offset, im->data->offset + im->data->size, im->data->path);
			}

			lseek_checked(fd, off_c);
			write_checked(fd, &bc, sizeof bc);

			printf("\n");
		} else if (files_written < n_files) {
			printf("file list block %d\n\n", i);

			struct bootfs_files bf = {
				.magic = ":BFILES:",
			};

			int files_rem = n_files - files_written;
			files_rem = files_rem > 51 ? 51 : files_rem;

			for (int i = 0; i < files_rem; i++) {
				struct file* f = &files[files_written];

				snprintf(bf.files[i].name, sizeof bf.files[i].name, "%s", f->name);
				bf.files[i].data_off = htobe32(f->data->offset);
				bf.files[i].data_len = htobe32(f->data->size);

				printf("  %08x-%08x %s %s\n",
				       f->data->offset, f->data->offset + f->data->size,
				       f->name, f->data->path);

				files_written++;
			}

			lseek_checked(fd, off_c);
			write_checked(fd, &bf, sizeof bf);

			printf("\n");
		} else {
			char zero[2048] = {0};
			lseek_checked(fd, off_c);
			write_checked(fd, &zero, sizeof zero);
		}

		off_c += 2048;
	}

	printf("Total filesystem size %zu KiB\n\n", off_i / 1024);

	close(fd);
	sync();
	return 0;
}
